MetadataMigration/src/test/java/org/opensearch/migrations/MultiTypeMappingTransformationTest.java

```diff
@@ -0,0 +1,301 @@
+
+/**
+ * Test class to verify custom transformations during metadata migrations.
+ */
+@Tag("isolatedTest")
+@Slf4j
+class MultiTypeMappingTransformationTest {
+
+    @TempDir
+    private File localDirectory;
+
+    @SneakyThrows
+    @Test
+    public void multiTypeTransformationTest_union() {
+        try (
+                final SearchClusterContainer indexCreatedCluster = new SearchClusterContainer(SearchClusterContainer.ES_V5_6_13);
+                final SearchClusterContainer upgradedSourceCluster = new SearchClusterContainer(SearchClusterContainer.ES_V6_8_23)
+                        .withFileSystemBind(localDirectory.getAbsolutePath(), SearchClusterContainer.CLUSTER_SNAPSHOT_DIR, BindMode.READ_WRITE);
+                final SearchClusterContainer targetCluster = new SearchClusterContainer(SearchClusterContainer.OS_V2_14_0)) {
+            CompletableFuture.allOf(
+                    CompletableFuture.runAsync(indexCreatedCluster::start),
+                    CompletableFuture.runAsync(upgradedSourceCluster::start),
+                    CompletableFuture.runAsync(targetCluster::start)
+            ).join();
+
+            var indexCreatedOperations = new ClusterOperations(indexCreatedCluster.getUrl());
+            var upgradedSourceOperations = new ClusterOperations(upgradedSourceCluster.getUrl());
+            var targetOperations = new ClusterOperations(targetCluster.getUrl());
+
+            // Test data
+            var originalIndexName = "test_index";
+
+            // Create index and add a document on the source cluster
+            indexCreatedOperations.createIndex(originalIndexName);
+            indexCreatedOperations.createDocument(originalIndexName, "1", "{\"field1\":\"My Name\"}", null, "type1");
+            indexCreatedOperations.createDocument(originalIndexName, "2", "{\"field2\":123}", null, "type2");
+            indexCreatedOperations.createDocument(originalIndexName, "3", "{\"field3\":1.1}", null, "type3");
+
+            var arguments = new MigrateOrEvaluateArgs();
+
+            // Use SnapshotImage as the transfer medium
+            var snapshotName = "initial-setup-snapshot";
+            var snapshotContext = SnapshotTestContext.factory().noOtelTracking();
+            var sourceClient = new OpenSearchClient(ConnectionContextTestParams.builder()
+                    .host(indexCreatedCluster.getUrl())
+                    .insecure(true)
+                    .build()
+                    .toConnectionContext());
+            var snapshotCreator = new FileSystemSnapshotCreator(
+                    snapshotName,
+                    sourceClient,
+                    SearchClusterContainer.CLUSTER_SNAPSHOT_DIR,
+                    List.of(),
+                    snapshotContext.createSnapshotCreateContext()
+            );
+
+            // Get snapshot for ES 5
+            SnapshotRunner.runAndWaitForCompletion(snapshotCreator);
```

Comment:
Refactor all of the ES 5 & snapshot creation into a function to keep the flow cleaner.